src/graphql/graphql-api.php

```diff
@@ -0,0 +1,255 @@
+			);
+		}
+
+		return $block;
+	}
+
+	/**
+	 * Flatten the inner blocks, no matter how many levels of nesting is there.
+	 *
+	 * @param array  $inner_blocks the inner blocks in the block.
+	 * @param string $parent_id ID of the parent block, that the inner blocks belong to.
+	 * @param array  $flattened_blocks the flattened blocks that's built up as we go through the inner blocks.
+	 * 
+	 * @return array
+	 */
+	public static function flatten_inner_blocks( $inner_blocks, $parent_id, $flattened_blocks = [] ) {
+		foreach ( $inner_blocks as $inner_block ) {
+				// Set the parentId to be the ID of the parent block whose inner blocks are being flattened.
+				$inner_block['parentId'] = $parent_id;
+
+				// This block doesnt have any inner blocks, so just add it to the flattened blocks. Ensure the parentId is set.
+			if ( ! isset( $inner_block['innerBlocks'] ) ) {
+				array_push( $flattened_blocks, $inner_block );
+				// This block is has inner blocks, so go through the inner blocks recursively.
+			} else {
+				$inner_blocks_copy = $inner_block['innerBlocks'];
+				unset( $inner_block['innerBlocks'] );
+				// first add the current block to the flattened blocks, and then go through the inner blocks recursively.
+				array_push( $flattened_blocks, $inner_block );
+				$flattened_blocks = self::flatten_inner_blocks( $inner_blocks_copy, $inner_block['id'], $flattened_blocks );
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		/**
+		 * Filter to enable/disable the graphQL API. By default, it is enabled.
+		 * 
+		 * @param bool $is_graphql_to_be_enabled Whether the graphQL API should be enabled or not.
+		 */
+		$is_graphql_to_be_enabled = apply_filters( 'vip_block_data_api__is_graphql_enabled', true );
+
+		if ( ! $is_graphql_to_be_enabled ) {
+			return;
+		}
+
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockAttribute',
+			[
+				'description' => 'Block attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => [ 'non_null' => 'String' ],
+						'description' => 'Block data attribute name',
+					],
+					'value' => [
+						'type'        => [ 'non_null' => 'String' ],
+						'description' => 'Block data attribute value',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual block, with the above attribute.
+		register_graphql_type(
+			'BlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'          => [
+						'type'        => [ 'non_null' => 'ID' ],
+						'description' => 'ID of the block',
+					],
+					'parentId'    => [
+						'type'        => 'ID',
+						'description' => 'ID of the parent for this inner block, if it is an inner block. Otherwise, it will not be set. If it set, this will match the ID of the block',
```

Comment:
last part confusing / redundant
```suggestion
						'description' => 'ID of the parent for this inner block, if it is an inner block. Otherwise, it will be null.',
```